### PR TITLE
Add check to validate our semver compliance

### PR DESCRIPTION
**Pull request by SamBarker**
### Type of change

- Enhancement / new feature

### Description

Adds a plug-in to fail the build on version changes which fail semver compatibility. I.e. if we made a breaking API change in a minor release.

as an example 
```
[INFO] --- japicmp-maven-plugin:0.18.0:cmp (default) @ kroxylicious-api ---
[INFO] Written file '/Users/sbarker/development/kroxy/kroxylicious/kroxylicious-api/target/japicmp/japicmp.diff'.
[INFO] Written file '/Users/sbarker/development/kroxy/kroxylicious/kroxylicious-api/target/japicmp/japicmp.xml'.
[INFO] Written file '/Users/sbarker/development/kroxy/kroxylicious/kroxylicious-api/target/japicmp/japicmp.html'.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  14.669 s
[INFO] Finished at: 2023-09-29T14:46:48+13:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal com.github.siom79.japicmp:japicmp-maven-plugin:0.18.0:cmp (default) on project kroxylicious-api: There is at least one incompatibility: io.kroxylicious.proxy.filter.FilterContext.sendRequest(short,org.apache.kafka.common.protocol.ApiMessage):CLASS_GENERIC_TEMPLATE_CHANGED,io.kroxylicious.proxy.service.BaseContributor.builder():CLASS_GENERIC_TEMPLATE_CHANGED,io.kroxylicious.proxy.service.BaseContributor.BaseContributor(io.kroxylicious.proxy.service.BaseContributor$BaseContributorBuilder):CONSTRUCTOR_LESS_ACCESSIBLE,io.kroxylicious.proxy.service.BaseContributor:CLASS_GENERIC_TEMPLATE_CHANGED,io.kroxylicious.proxy.service.BaseContributor$BaseContributorBuilder.add(java.lang.String,java.lang.Class,java.util.function.BiFunction,boolean):CLASS_GENERIC_TEMPLATE_CHANGED,io.kroxylicious.proxy.service.BaseContributor$BaseContributorBuilder:CLASS_GENERIC_TEMPLATE_CHANGED,io.kroxylicious.proxy.service.Contributor.contributes(java.lang.String):METHOD_ADDED_TO_INTERFACE,io.kroxylicious.proxy.service.Contributor.getConfigDefinition(java.lang.String):METHOD_ADDED_TO_INTERFACE,io.kroxylicious.proxy.service.Contributor.getInstance(java.lang.String,io.kroxylicious.proxy.config.BaseConfig):METHOD_REMOVED,io.kroxylicious.proxy.service.Contributor.getInstance(java.lang.String,io.kroxylicious.proxy.service.Context):METHOD_ADDED_TO_INTERFACE,io.kroxylicious.proxy.service.Contributor:CLASS_GENERIC_TEMPLATE_CHANGED -> [Help 1]
```

### Additional Context

To aid the discussion about API maturity annotations.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Review performance test results. Ensure that any degradations to performance numbers are understood and justified.
- [ ] Make sure all Sonar-Lint warnings are addressed or are justifiably ignored.
- [ ] Update documentation
- [ ] Reference relevant issue(s) and close them after merging
- [ ] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

### CODE REVIEW

File: kroxylicious-api/pom.xml [hunk 1]
```diff
@@ -174,6 +174,34 @@
...
+                        <dependency>
+                            <groupId>${project.groupId}</groupId>
+                            <artifactId>${project.artifactId}</artifactId>
+                            <version>0.2.0</version>
+                            <type>jar</type>
+                        </dependency>
+                    </oldVersion>
+                    <parameter>
+                        <breakBuildOnBinaryIncompatibleModifications>true</breakBuildOnBinaryIncompatibleModifications>
+                        <breakBuildBasedOnSemanticVersioningForMajorVersionZero>true</breakBuildBasedOnSemanticVersioningForMajorVersionZero>
```
Comment (SamBarker) [inline kroxylicious-api/pom.xml:192]: set to force a failure in this PR as an example. I'm not advocating any particular settings at this stage.

**Comment by k-wall** [comment]
Looks like a great idea, let's try it.

File: kroxylicious-api/pom.xml [hunk 2]
```diff
@@ -174,6 +174,34 @@
...
+            <plugin>
+                <groupId>com.github.siom79.japicmp</groupId>
+                <artifactId>japicmp-maven-plugin</artifactId>
+                <version>0.18.0</version>
+                <configuration>
+                    <oldVersion>
+                        <dependency>
+                            <groupId>${project.groupId}</groupId>
+                            <artifactId>${project.artifactId}</artifactId>
+                            <version>0.2.0</version>
```
[the remark this answers is not in the document]
Comment (SamBarker) [reply kroxylicious-api/pom.xml:186]: I'm hoping to get it updated automatically in the release process but yes if we use this it should be a property.

From the docs I don't think this should be required. The mojo should be able to work it out. However in my quick poc that didn't seem to work hence I added it.

**Comment by tombentley** [comment]
@SamBarker are you planning to get back to this?

**Comment by SamBarker** [comment]
Yeah I think it's probably worth doing

**Pull request closed without merging**